fix(docs): updating docs structure

critesjosh opened the pull request:
:warning: this PR builds off of the changes in https://github.com/AztecProtocol/aztec-packages/pull/16442, so merge that one first

This PR makes several edits to the docs. The main impetus behind the changes being made is to make a more linear learning journey for developers as they move down the sidebar.

- moves the guides section above the tutorials
- flattens the pages inside the "Developer smart contracts" section
- makes the "developing smart contracts" section a more linear flow for how a developer should be moving through this content
- moves some of the technical explanations to the concepts section
- adds a several pages to the Guides section including:
  - project sturcturing
  - data types
  - custom types
  - defining functions (initializer page moved into here)
 - removes some explainers and reference about storage from the references section into the guides section

closes: https://github.com/AztecProtocol/dev-rel/issues/598

signorecello reviewed (approved):
LGTM. I could look for small nits but I'd rather iterate fast than dragging this PR

AztecBot committed:
fix(docs): updating docs structure

:warning: this PR builds off of the changes in https://github.com/AztecProtocol/aztec-packages/pull/16442, so merge that one first

This PR makes several edits to the docs. The main impetus behind the changes being made is to make a more linear learning journey for developers as they move down the sidebar.

- moves the guides section above the tutorials
- flattens the pages inside the "Developer smart contracts" section
- makes the "developing smart contracts" section a more linear flow for how a developer should be moving through this content
- moves some of the technical explanations to the concepts section
- adds a several pages to the Guides section including:
  - project sturcturing
  - data types
  - custom types
  - defining functions (initializer page moved into here)
 - removes some explainers and reference about storage from the references section into the guides section

closes: https://github.com/AztecProtocol/dev-rel/issues/598

Co-authored-by: Alex Gherghisan <alexghr@users.noreply.github.com>
Co-authored-by: Charlie Lye <5764343+charlielye@users.noreply.github.com>
Co-authored-by: Facundo <fcarreiro@users.noreply.github.com>
Co-authored-by: IlyasRidhuan <ilyasridhuan@gmail.com>
Co-authored-by: Jonathan Hao <jonathanpohsianghao@gmail.com>
Co-authored-by: Josh Crites <jc@joshcrites.com>
Co-authored-by: José Pedro Sousa <outgoing@zpedro.dev>
Co-authored-by: LHerskind <16536249+LHerskind@users.noreply.github.com>
Co-authored-by: Lasse Herskind <16536249+LHerskind@users.noreply.github.com>
Co-authored-by: Raju Krishnamoorthy <krishnamoorthy@gmail.com>
Co-authored-by: fcarreiro <facundo@aztecprotocol.com>
Co-authored-by: federicobarbacovi <171914500+federicobarbacovi@users.noreply.github.com>
Co-authored-by: jeanmon <jean@aztec-labs.com>
Co-authored-by: maramihali <mara@aztec-labs.com>
Co-authored-by: sergei iakovenko <105737703+iakovenkos@users.noreply.github.com>
Co-authored-by: sirasistant <sirasistant@gmail.com>
Co-authored-by: thunkar <gregojquiros@gmail.com>
